File: crates/zensical/src/structure/markdown/autorefs.rs
// Copyright (c) 2025 Zensical and contributors

// SPDX-License-Identifier: MIT
// Third-party contributions licensed under DCO

// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to
// deal in the Software without restriction, including without limitation the
// rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
// sell copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:

// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.

// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
// IN THE SOFTWARE.

// ----------------------------------------------------------------------------

//! Autorefs (mkdocstrings).

use ahash::HashMap;
use pyo3::FromPyObject;
use regex::{Captures, Regex};
use serde::{Deserialize, Serialize};
use std::path::Path;
use std::string::ToString;
use std::sync::LazyLock;
use zrx::path::PathExt;

// ----------------------------------------------------------------------------
// Constants
// ----------------------------------------------------------------------------

/// Autoref regex.
static AUTOREF_RE: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"<autoref (?P<attrs>.*?)>(?P<title>.*?)</autoref>").unwrap()
});

/// Handled autoref attributes that should not be passed through to the output link.
const HANDLED_ATTRS: &[&str] = &[
    "identifier",
    "optional",
    "hover",
    "class",
    "domain",
    "role",
    "origin",
    "filepath",
    "lineno",
    "slug",
    "backlink-type",
    "backlink-anchor",
];

// ----------------------------------------------------------------------------
// Helper Functions
// ----------------------------------------------------------------------------

/// Escapes HTML special characters.
fn html_escape(text: &str) -> String {
    text.replace('&', "&amp;")
        .replace('<', "&lt;")
        .replace('>', "&gt;")
        .replace('"', "&quot;")
        .replace('\'', "&#x27;")
}

/// Helper to check if a URL is relative to a base URL.
fn is_relative_to(url: &str, base: &str) -> bool {
    // Remove fragments and query strings for directory comparison
    let url_path = url
        .split('#')
        .next()
        .unwrap_or(url)
        .split('?')
        .next()
        .unwrap_or(url);
    let base_path = base
        .split('#')
        .next()
        .unwrap_or(base)
        .split('?')
        .next()
        .unwrap_or(base);

    // Use Path::starts_with for proper path comparison
    Path::new(url_path).starts_with(Path::new(base_path))
}

/// Gets the parent path of a URL.
fn parent_path(url: &str) -> Option<String> {
    Path::new(url)
        .parent()
        .and_then(|p| p.to_str())
        .map(ToString::to_string)
}

/// Resolves the closest URL from a list relative to from_url.
///
/// We do that when multiple URLs are found for an identifier.
///
/// By closest, we mean a combination of "relative to the current page" and "shortest distance from the current page".
///
/// For example, if you link to identifier `hello` from page `foo/bar/`,
/// and the identifier is found in `foo/`, `foo/baz/` and `foo/bar/baz/qux/` pages,
/// autorefs will resolve to `foo/bar/baz/qux`, which is the only URL relative to `foo/bar/`.
///
/// If multiple URLs are equally close, autorefs will resolve to the first of these equally close URLs.
/// If autorefs cannot find any URL that is close to the current page, it will log a warning and resolve to the first URL found.
///
/// When false and multiple URLs are found for an identifier, autorefs will log a warning and resolve to the first URL.
fn resolve_closest_url(
    from_url: &str, urls: &[String], _qualifier: &str,
) -> String {
    let mut base_url = from_url.to_string();
    let candidates;

    loop {
        let found: Vec<String> = urls
            .iter()
            .filter(|url| is_relative_to(url, &base_url))
            .cloned()
            .collect();

        if !found.is_empty() {
            candidates = found;
            break;
        }

        match parent_path(&base_url) {
            Some(parent) if !parent.is_empty() => {
                base_url = parent;
            }
            _ => {
                // @todo Log warning using qualifier
                return urls[0].clone();
            }
        }
    }

    if candidates.len() == 1 {
        candidates[0].clone()
    } else {
        // Find the URL with the fewest slashes
        candidates
            .into_iter()
            .min_by_key(|url| url.matches('/').count())
            .unwrap()
    }
}

/// Computes a relative URL from from_url to to_url.
fn relative_url(from_url: &str, to_url: &str) -> String {
    let from_path = Path::new(from_url);

    // Split URL and fragment for relative computation
    let (to_path, to_fragment) = to_url
        .split_once('#')
        .map_or((Path::new(to_url), None), |(path, f)| {
            (Path::new(path), Some(f))
        });

    // Make target URL relative to page
    let mut rel_path = to_path
        .relative_to(from_path)
        .to_string_lossy()
        .replace('\\', "/");

    // Add fragment back if present
    if let Some(frag) = to_fragment {
        // If the relative path is "." and we have a fragment,
        // just return the fragment
        if rel_path == "." {
            return format!("#{frag}");
        }
        // If `to_path` was empty (URL was just a fragment),
        // add "/" before the fragment
        if to_path.as_os_str().is_empty() {
            rel_path.push('/');
        }
        rel_path.push('#');
        rel_path.push_str(frag);
    }

    rel_path
}

/// Checks if a URL is relative (no scheme).
fn is_relative_url(url: &str) -> bool {
    !(url.starts_with("http://") || url.starts_with("https://"))
}

// ----------------------------------------------------------------------------
// Structs
// ----------------------------------------------------------------------------

/// Autorefs (mkdocstrings).
///
/// We use three URL maps, one for "primary" URLs, one for "secondary" URLs,
/// and one for "absolute" URLs.
///
/// - A primary URL is an identifier that links to a specific anchor on a page.
/// - A secondary URL is an alias of an identifier that links to the same anchor as the identifier's primary URL.
///   Primary URLs with these aliases as identifiers may or may not be rendered later.
/// - An absolute URL is an identifier that links to an external resource.
///   These URLs are typically registered by mkdocstrings when loading object inventories.
///
/// mkdocstrings registers a primary URL for each heading rendered in a page.
/// Then, for each alias of this heading's identifier, it registers a secondary URL.
///
/// For example:
///
/// - Object `a.b.c.d` has aliases `a.b.d` and `a.d`
/// - Object `a.b.c.d` is rendered.
/// - We register `a.b.c.d` -> page#a.b.c.d as primary
/// - We register `a.b.d` -> page#a.b.c.d as secondary
/// - We register `a.d` -> page#a.b.c.d as secondary
/// - Later, if `a.b.d` or `a.d` are rendered, we will register primary and secondary URLs the same way
/// - This way we are sure that each of `a.b.c.d`, `a.b.d` or `a.d` will link to their primary URL, if any, or their secondary URL, accordingly
///
/// We need to keep track of whether an identifier is primary or secondary,
/// to give it precedence when resolving cross-references.
/// We wouldn't want to log a warning if there is a single primary URL and one or more secondary URLs,
/// instead we want to use the primary URL without any warning.
///
/// - A single primary URL mapped to an identifer? Use it.
/// - Multiple primary URLs mapped to an identifier? Use the first one, or closest one if configured as such.
/// - No primary URL mapped to an identifier, but a secondary URL mapped? Use it.
/// - Multiple secondary URLs mapped to an identifier? Use the first one, or closest one if configured as such.
/// - No secondary URL mapped to an identifier? Try using absolute URLs
///   (typically registered by loading inventories in mkdocstrings).
#[derive(
    Clone, Debug, Default, FromPyObject, Serialize, Deserialize, PartialEq, Eq,
)]
#[pyo3(from_item_all)]
pub struct Autorefs {
    // Primary URLs.
    pub primary: HashMap<String, Vec<String>>,
    // Secondary URLs.
    pub secondary: HashMap<String, Vec<String>>,
    // Inventory URLs.
    pub inventory: HashMap<String, String>,
    // Titles.
    pub titles: HashMap<String, String>,
}

// ----------------------------------------------------------------------------
// Implementations
// ----------------------------------------------------------------------------

impl Autorefs {
    /// Creates a new, empty autorefs.
    pub fn new() -> Self {
        Self::default()
    }

    /// Parses HTML attributes string into a HashMap.
    ///
    /// @todo Document that this is not the most resilient HTML parser
    /// but since we control the autorefs elements, it's fine for now
    fn parse_attributes(attrs_str: &str) -> HashMap<String, String> {
        let mut attrs = HashMap::default();
        let mut chars = attrs_str.chars().peekable();

        while let Some(ch) = chars.peek() {
            // Skip whitespace
            if ch.is_whitespace() {
                chars.next();
                continue;
            }

            // Parse attribute name
            let mut name = String::new();
            while let Some(&ch) = chars.peek() {
                if ch.is_whitespace() || ch == '=' {
                    break;
                }
                name.push(ch);
                chars.next();
            }

            if name.is_empty() {
                break;
            }

            // Skip whitespace
            while let Some(&ch) = chars.peek() {
                if !ch.is_whitespace() {
                    break;
                }
                chars.next();
            }

            // Check for '='
            let has_value = chars.peek() == Some(&'=');
            if has_value {
                chars.next(); // consume '='

                // Skip whitespace after '='
                while let Some(&ch) = chars.peek() {
                    if !ch.is_whitespace() {
                        break;
                    }
                    chars.next();
                }

                // Parse value
                let value = if let Some(&quote) = chars.peek() {
                    if quote == '"' || quote == '\'' {
                        chars.next(); // consume opening quote
                        let mut val = String::new();
                        for ch in chars.by_ref() {
                            if ch == quote {
                                break; // consume closing quote
                            }
                            val.push(ch);
                        }
                        val
                    } else {
                        // Unquoted value
                        let mut val = String::new();
                        while let Some(&ch) = chars.peek() {
                            if ch.is_whitespace() {
                                break;
                            }
                            val.push(ch);
                            chars.next();
                        }
                        val
                    }
                } else {
                    String::new()
                };

                attrs.insert(name, value);
            } else {
                // Boolean attribute
                attrs.insert(name, String::new());
            }
        }

        attrs
    }

    /// Resolves the URL for an item identifier (internal implementation).
    fn get_url_from_id(
        &self, identifier: &str, from_url: &str, resolve_closest: bool,
    ) -> Result<String, String> {
        // Try primary URLs first - usually, an object should not have multiple
        // primary URLs, but if it does, resolve closest if requested. Primary
        // URLs are the canonical locations objects are defined. If an object
        // is re-exported, it should have a secondary URL instead.
        if let Some(urls) = self.primary.get(identifier) {
            if urls.len() > 1 && resolve_closest {
                return Ok(resolve_closest_url(from_url, urls, "primary"));
                // @todo Log warning about multiple URLs in production
            }
            return Ok(urls[0].clone());
        }

        // Try secondary URLs
        if let Some(urls) = self.secondary.get(identifier) {
            if urls.len() > 1 {
                // Always resolve closest for secondary
                //
                // Downstream projects rendering aliases of objects
                // imported from upstream ones will render these upstream
                // objects' docstrings. These docstrings can contain
                // cross-references to other upstream objects that are not
                // rendered directly in downstream project's docs.
                //
                // If downstream project renders subclasses of upstream
                // class, with inherited members, only primary URLs will be
                // registered for the aliased/downstream identifiers, and
                // only secondary URLs will be registered for the upstream
                // identifiers.
                //
                // When trying to apply the cross-reference
                // for the upstream docstring, autorefs will find only
                // secondary URLs, and multiple ones. But the end user does
                // not have control over this. It means we shouldn't log
                // warnings when multiple secondary URLs are found, and
                // always resolve to closest.
                return Ok(resolve_closest_url(from_url, urls, "secondary"));
            }
            return Ok(urls[0].clone());
        }

        // Try inventory (absolute URLs)
        if let Some(url) = self.inventory.get(identifier) {
            return Ok(url.clone());
        }

        Err(format!("Identifier '{identifier}' not found"))
    }

    /// Gets the URL for an item identifier.
    fn get_url_and_title_from_id(
        &self, identifier: &str, from_url: &str,
    ) -> Result<(String, Option<String>), String> {
        let mut url = self.get_url_from_id(identifier, from_url, true)?;

        // Get title using URL as key (not identifier)
        let title = self.titles.get(&url).cloned();

        // If from_url is provided and URL is relative, compute relative URL
        if is_relative_url(&url) {
            url = relative_url(from_url, &url);
        }

        Ok((url, title))
    }

    /// Resolves the URL for the first matching identifier.
    fn get_url_and_title_from_ids(
        &self, identifiers: &[String], from_url: &str,
    ) -> Result<(String, Option<String>), String> {
        for identifier in identifiers {
            if let Ok(result) =
                self.get_url_and_title_from_id(identifier, from_url)
            {
                return Ok(result);
            }
        }
        Err(format!(
            "None of the identifiers {identifiers:?} were found",
        ))
    }

    /// Extends autorefs with another instance.
    pub fn extend(&mut self, other: Autorefs) {
        for (key, values) in other.primary {
            self.primary.entry(key).or_default().extend(values);
        }
        for (key, values) in other.secondary {
            self.secondary.entry(key).or_default().extend(values);
        }
        for (key, value) in other.inventory {
            self.inventory.insert(key, value);
        }
        for (key, value) in other.titles {
            self.titles.insert(key, value);
        }
    }

    /// Replaces autorefs in the given content.
    pub fn replace_in(&self, content: String, from_url: &str) -> String {
        let output = AUTOREF_RE.replace_all(&content, |captures: &Captures| {
            let attrs_str =
                captures.name("attrs").map_or("", |m| m.as_str());
            let title =
                captures.name("title").map_or("", |m| m.as_str());

            // Parse the HTML attributes
            let attrs = Self::parse_attributes(attrs_str);
            let identifier =
                attrs.get("identifier").cloned().unwrap_or_default();
            let slug = attrs.get("slug").cloned().unwrap_or_default();
            let optional = attrs.contains_key("optional");

            let identifiers = if slug.is_empty() {
                vec![identifier.clone()]
            } else {
                vec![identifier.clone(), slug.clone()]
            };

            match self.get_url_and_title_from_ids(&identifiers, from_url) {
                Ok((url, original_title)) => {
                    // Check if URL is external (not relative)
                    let external = !is_relative_url(&url);

                    // Build CSS classes
                    let mut classes = vec![
                        "autorefs".to_string(),
                        if external {
                            "autorefs-external".to_string()
                        } else {
                            "autorefs-internal".to_string()
                        },
                    ];

                    // Add existing classes from attrs
                    if let Some(class_str) = attrs.get("class") {
                        classes.extend(
                            class_str
                                .split_whitespace()
                                .map(ToString::to_string),
                        );
                    }
                    let class_attr = classes.join(" ");

                    // Build remaining attributes (those not in the handled set)
                    let remaining_attrs: Vec<String> = attrs
                        .iter()
                        .filter(|(k, _)| !HANDLED_ATTRS.contains(&k.as_str()))
                        .map(|(k, v)| {
                            if v.is_empty() {
                                // Boolean attribute (no value)
                                k.clone()
                            } else {
                                // Attribute with value
                                format!("{k}=\"{v}\"")
                            }
                        })
                        .collect();

                    let remaining = if remaining_attrs.is_empty() {
                        String::new()
                    } else {
                        format!(" {}", remaining_attrs.join(" "))
                    };

                    // Build title attribute (link_titles is always true, strip_title_tags is always false)
                    let tooltip = if optional {
                        // For optional, we use identifier as fallback if no original_title
                        original_title.as_deref().unwrap_or(&identifier).to_string()
                    } else {
                        // For non-optional, use original_title or empty
                        original_title.as_deref().unwrap_or("").to_string()
                    };

                    let title_attr = if !tooltip.is_empty() && !format!("<code>{title}</code>").contains(&tooltip) {
                        format!(" title=\"{}\"", html_escape(&tooltip))
                    } else {
                        String::new()
                    };

                    let escaped_url = html_escape(&url);
                    format!(
                        "<a class=\"{class_attr}\"{title_attr} href=\"{escaped_url}\"{remaining}>{title}</a>"
                    )
                }
                Err(_) => {
                    if optional {
                        format!("<span title=\"{identifier}\">{title}</span>")
                    } else {
                        // @todo: unmapped.append((identifier, attrs.context))
                        if title == identifier {
                            format!("[{identifier}][]")
                        } else if title == format!("<code>{identifier}</code>")
                            && slug.is_empty()
                        {
                            format!("[<code>{identifier}</code>][]")
                        } else {
                            format!("[{title}][{identifier}]")
                        }
                    }
                }
            }
        });

        output.to_string()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_resolve_closest_url() {
        let test_cases = vec![
            ("", vec!["x/#b", "#b"], "#b"),
            ("a/b", vec!["x/#e", "a/c/#e", "a/d/#e"], "a/c/#e"),
            ("a/b/", vec!["x/#e", "a/d/#e", "a/c/#e"], "a/d/#e"),
            ("a/b", vec!["x/#e", "a/c/#e", "a/c/d/#e"], "a/c/#e"),
            ("a/b/", vec!["x/#e", "a/c/d/#e", "a/c/#e"], "a/c/#e"),
            (
                "a/b/c",
                vec!["x/#e", "a/#e", "a/b/#e", "a/b/c/#e", "a/b/c/d/#e"],
                "a/b/c/#e",
            ),
            (
                "a/b/c/",
                vec!["x/#e", "a/#e", "a/b/#e", "a/b/c/d/#e", "a/b/c/#e"],
                "a/b/c/#e",
            ),
            ("a", vec!["b/c/#d", "c/#d"], "b/c/#d"),
            ("a/", vec!["c/#d", "b/c/#d"], "c/#d"),
        ];

        for (base, urls, expected) in test_cases {
            let urls: Vec<String> =
                urls.into_iter().map(String::from).collect();
            let result = resolve_closest_url(base, &urls, "test");
            assert_eq!(result, expected, "Failed for base: {base}");
        }
    }

    #[test]
    fn test_relative_url() {
        let test_cases = vec![
            ("a/", "a#b", "#b"),
            ("a/", "a/b#c", "b#c"),
            ("a/b/", "a/b#c", "#c"),
            ("a/b/", "a/c#d", "../c#d"),
            ("a/b/", "a#c", "..#c"),
            ("a/b/c/", "d#e", "../../../d#e"),
            ("a/b/", "c/d/#e", "../../c/d/#e"),
            ("a/index.html", "a/index.html#b", "#b"),
            ("a/index.html", "a/b.html#c", "b.html#c"),
            ("a/b.html", "a/b.html#c", "#c"),
            ("a/b.html", "a/c.html#d", "c.html#d"),
            ("a/b.html", "a/index.html#c", "index.html#c"),
            ("a/b/c.html", "d.html#e", "../../d.html#e"),
            ("a/b.html", "c/d.html#e", "../c/d.html#e"),
            ("a/b/index.html", "a/b/c/d.html#e", "c/d.html#e"),
            ("", "#x", "#x"),
            ("a/", "#x", "../#x"),
            ("a/b.html", "#x", "../#x"),
            ("", "a/#x", "a/#x"),
            ("", "a/b.html#x", "a/b.html#x"),
        ];

        for (current_url, to_url, expected_href) in test_cases {
            let result = relative_url(current_url, to_url);
            assert_eq!(
                result, expected_href,
                "Failed for relative_url('{current_url}', '{to_url}'), expected '{expected_href}' but got '{result}'"
            );
        }
    }
}

File: python/zensical/extensions/autorefs.py
# Copyright (c) 2025-2026 Zensical and contributors

# SPDX-License-Identifier: MIT
# All contributions are certified under the DCO

# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to
# deal in the Software without restriction, including without limitation the
# rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
# sell copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:

# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.

# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
# IN THE SOFTWARE.

from __future__ import annotations

import re
from abc import ABC, abstractmethod
from dataclasses import dataclass
from html import escape
from typing import TYPE_CHECKING, Any, TypeAlias
from xml.etree.ElementTree import Comment, Element

from markdown.core import Markdown
from markdown.extensions import Extension
from markdown.extensions.toc import slugify
from markdown.inlinepatterns import (
    REFERENCE_RE,
    ReferenceInlineProcessor,
)
from markdown.treeprocessors import Treeprocessor
from markdown.util import HTML_PLACEHOLDER_RE, INLINE_PLACEHOLDER_RE
from markupsafe import Markup

if TYPE_CHECKING:
    from pathlib import Path
    from re import Match

    from markdown import Markdown

    from zensical.extensions.context import Page


# ----------------------------------------------------------------------------
# Constants
# ----------------------------------------------------------------------------


# URL order determines which target wins when several are equally
# suitable. Dict keys give us that order plus constant-time lookups.
OrderedSet: TypeAlias = dict[str, None]
URLMap: TypeAlias = dict[str, OrderedSet]


# ----------------------------------------------------------------------------
# Globals
# ----------------------------------------------------------------------------


AUTOREFS: AutorefsStore | None = None


# ----------------------------------------------------------------------------
# Classes
# ----------------------------------------------------------------------------


class AutorefsStore:
    """Mock the autorefs plugin (data store)."""

    def __init__(self) -> None:
        self.current_page: Page | None = None
        self.scan_toc: bool = True
        self.record_backlinks: bool = False

        self._primary_url_map: URLMap = {}
        self._secondary_url_map: URLMap = {}
        self._abs_url_map: dict[str, str] = {}
        self._title_map: dict[str, str] = {}
        self._page_registrations: dict[str, set[tuple[bool, str, str]]] = {}

    def set_page(self, page: Page) -> None:
        """Set the current page and discard its previous registrations."""
        self.current_page = page
        self.pop_page(page.url)

    def pop_page(self, page_url: str) -> dict[str, Any]:
        """Remove and return registrations owned by one page."""
        registrations = self._page_registrations.pop(page_url, set())
        primary = self._pop_urls(self._primary_url_map, registrations, True)
        secondary = self._pop_urls(
            self._secondary_url_map, registrations, False
        )
        urls = {
            url
            for values in (primary, secondary)
            for entries in values.values()
            for url in entries
        }
        titles = {
            url: self._title_map.pop(url)
            for url in urls
            if url in self._title_map
        }
        return {
            "primary": primary,
            "secondary": secondary,
            "titles": titles,
        }

    @staticmethod
    def _pop_urls(
        url_map: URLMap,
        registrations: set[tuple[bool, str, str]],
        primary: bool,
    ) -> dict[str, list[str]]:
        """Remove registered URLs from one URL map, preserving their order."""
        selected: dict[str, set[str]] = {}
        for is_primary, identifier, url in registrations:
            if is_primary == primary:
                selected.setdefault(identifier, set()).add(url)

        result: dict[str, list[str]] = {}
        for identifier, selected_urls in selected.items():
            urls = url_map.get(identifier, {})
            result[identifier] = [url for url in urls if url in selected_urls]
            for url in selected_urls:
                urls.pop(url, None)
            if not urls:
                url_map.pop(identifier, None)
        return result

    def register_anchor(
        self,
        page: Page,
        identifier: str,
        anchor: str | None = None,
        *,
        title: str | None = None,
        primary: bool = True,
    ) -> None:
        url = f"{page.url}#{anchor or identifier}"
        url_map = self._primary_url_map if primary else self._secondary_url_map
        url_map.setdefault(identifier, {})[url] = None
        self._page_registrations.setdefault(page.url, set()).add(
            (primary, identifier, url)
        )
        if title and url not in self._title_map:
            self._title_map[url] = title

    def register_url(self, identifier: str, url: str) -> None:
        self._abs_url_map[identifier] = url


# Unusued yet, only when/if we vendor mkdocstrings and handlers
class AutorefsHookInterface(ABC):
    """An interface for hooking into how AutoRef handles inline references."""

    @dataclass
    class Context:
        """The context around an auto-reference."""

        domain: str
        role: str
        origin: str
        filepath: str | Path
        lineno: int

        def as_dict(self) -> dict[str, str]:
            """Convert the context to a dictionary of HTML attributes."""
            return {
                "domain": self.domain,
                "role": self.role,
                "origin": self.origin,
                "filepath": str(self.filepath),
                "lineno": str(self.lineno),
            }

    @abstractmethod
    def expand_identifier(self, identifier: str) -> str:
        """Expand an identifier in a given context."""
        raise NotImplementedError

    @abstractmethod
    def get_context(self) -> AutorefsHookInterface.Context:
        """Get the current context."""
        raise NotImplementedError


class AutorefsInlineProcessor(ReferenceInlineProcessor):
    """A Markdown extension to handle inline references."""

    # We have to keep the legacy name `mkdocs-autorefs` because mkdocstrings
    # checks that the name of the original inline processor is a member of
    # `self.md.inlinepatterns` to know whether it must assign the hook.
    # This can be changed to something else if we update mkdocstrings.
    name = "mkdocs-autorefs"
    hook: AutorefsHookInterface | None = None
    backlink_marker = "data-zensical-autoref"

    def __init__(
        self,
        *args: Any,
        record_backlinks: bool = False,
        **kwargs: Any,
    ) -> None:
        super().__init__(REFERENCE_RE, *args, **kwargs)
        self._record_backlinks = record_backlinks

    @property
    def stashed_nodes(self) -> dict[str, Element | str]:
        return self.md.treeprocessors["inline"].stashed_nodes

    def handleMatch(
        self, m: Match[str], data: str
    ) -> tuple[Element | None, int | None, int | None]:
        """Handle an element that matched."""
        text, index, handled = self.getText(data, m.end(0))
        if not handled:
            return None, None, None

        identifier, slug, end, handled = self._eval_id(data, index, text)
        if not handled or identifier is None:
            return None, None, None

        if slug is None and re.search(r"[\x00-\x1f]", identifier):
            # Do nothing if the matched reference still contains control
            # characters (from 0 to 31 included) that weren't unstashed when
            # trying to compute a slug of the title.
            return None, m.start(0), end

        return self._make_tag(identifier, text, slug=slug), m.start(0), end

    def _unstash(self, identifier: str) -> str:
        stashed_nodes = self.stashed_nodes

        def _repl(match: Match) -> str:
            el = stashed_nodes.get(match[1])
            if isinstance(el, Element):
                return f"`{''.join(el.itertext())}`"
            if el == "\x0296\x03":
                return "`"
            return str(el)

        return INLINE_PLACEHOLDER_RE.sub(_repl, identifier)

    def _eval_id(
        self, data: str, index: int, text: str
    ) -> tuple[str | None, str | None, int, bool]:
        """Evaluate the id portion of `[ref][id]`.

        If `[ref][]` use `[ref]`.
        """
        m = self.RE_LINK.match(data, pos=index)
        if not m:
            return None, None, index, False

        # Default; an identifier was provided, match it exactly (later).
        slug = None

        if not (identifier := m.group(1)):
            # Only a title was provided, use it as identifier.
            identifier = text

            # Catch single stash entries, like the result of [`Foo`][].
            if match := INLINE_PLACEHOLDER_RE.fullmatch(identifier):
                stashed_nodes = self.stashed_nodes
                el = stashed_nodes.get(match[1])
                if isinstance(el, Element) and el.tag == "code":
                    # The title was wrapped in backticks, we only keep the
                    # content and tell autorefs to match the identifier exactly.
                    identifier = "".join(el.itertext())
                    # Special case: allow pymdownx.inlinehilite raw <code>
                    # snippets but strip them back to unhighlighted.
                    if match := HTML_PLACEHOLDER_RE.fullmatch(identifier):
                        stash_index = int(match.group(1))
                        html = self.md.htmlStash.rawHtmlBlocks[stash_index]
                        identifier = Markup(html).striptags()  # noqa: S704
                        self.md.htmlStash.rawHtmlBlocks[stash_index] = escape(
                            identifier
                        )

            # In any other case, unstash the title and slugify it.
            # Examples: ``[`Foo` and `Bar`]``, `[The *Foo*][]`.
            else:
                identifier = self._unstash(identifier)
                slug = slugify(identifier, separator="-")

        end = m.end(0)
        return identifier, slug, end, True

    def _make_tag(
        self, identifier: str, text: str, *, slug: str | None = None
    ) -> Element:
        """Create a tag that can be resolved after site settlement."""
        el = Element("autoref")
        if self._record_backlinks:
            el.set(self.backlink_marker, "")
        if self.hook:
            identifier = self.hook.expand_identifier(identifier)
            el.attrib.update(self.hook.get_context().as_dict())
        el.set("identifier", identifier)
        el.text = text
        if slug:
            el.attrib["slug"] = slug
        return el


class BacklinkContextTreeProcessor(Treeprocessor):
    """Expose one Markdown conversion's backlink context to the Rust visitor.

    Mkdocstrings mutates ``initial_id`` around nested Markdown conversions. The
    two constant-time boundary markers preserve that context without walking
    the ElementTree in Python.
    """

    name = "mkdocs-autorefs-backlinks"
    initial_id: str | None = None

    def run(self, root: Element) -> None:
        """Bracket this rendered fragment with its initial heading ID."""
        start = "zensical:autoref-context:start"
        if self.initial_id:
            start = f"{start}:{self.initial_id.encode().hex()}"
        root.insert(0, Comment(start))
        root.append(Comment("zensical:autoref-context:end"))


class AutorefsExtension(Extension):
    """Extension that transforms unresolved references into auto-references.

    Auto-references are resolved later, on the Rust side.
    """

    name = "zensical.extensions.autorefs"

    def __init__(self, **kwargs: Any) -> None:
        """Initialize the extension."""
        self._enabled: bool = kwargs.pop("enabled", True)
        self._record_backlinks: bool = kwargs.pop("record_backlinks", False)

    def extendMarkdown(self, md: Markdown) -> None:
        """Register the Markdown extension."""
        if not self._enabled:
            return
        md.registerExtension(self)

        record_backlinks = self._record_backlinks and (
            "attr_list" in md.treeprocessors or "toc" in md.treeprocessors
        )
        inline_processor = AutorefsInlineProcessor(
            md,
            record_backlinks=record_backlinks,
        )
        md.inlinePatterns.register(
            inline_processor,
            AutorefsInlineProcessor.name,
            168,  # after markdown.inlinepatterns.ReferenceInlineProcessor
        )

        if record_backlinks:
            store = get_autorefs_store()
            store.record_backlinks = True
            processor = BacklinkContextTreeProcessor(md)
            md.treeprocessors.register(processor, processor.name, 0)


# ----------------------------------------------------------------------------
# Functions
# ----------------------------------------------------------------------------


def get_autorefs_store() -> AutorefsStore:
    """Get the global autorefs instance."""
    global AUTOREFS  # noqa: PLW0603
    if AUTOREFS is None:
        AUTOREFS = AutorefsStore()
    return AUTOREFS


def get_autorefs_page_data(page_url: str) -> dict[str, Any]:
    """Take page-local autorefs data.

    Rust combines these registrations into the settled URL registry used to
    resolve `<autoref>` elements emitted by autorefs and mkdocstrings.
    """
    if AUTOREFS:
        return AUTOREFS.pop_page(page_url)
    return {"primary": {}, "secondary": {}, "titles": {}}


def get_autorefs_inventory_data() -> dict[str, str] | None:
    """Return global inventory URLs if Markdown rendering initialized them."""
    if AUTOREFS is None:
        return None
    return AUTOREFS._abs_url_map


def set_autorefs_page(page: Page) -> None:
    """Set autorefs current page."""
    store = get_autorefs_store()
    store.set_page(page)


def reset() -> None:
    """Reset global state in-between rebuilds."""
    global AUTOREFS  # noqa: PLW0603
    AUTOREFS = None


def makeExtension(**kwargs: Any) -> AutorefsExtension:
    """Register Markdown extension."""
    return AutorefsExtension(**kwargs)
